lib/Horde/Auth/Dummy.php

```diff
@@ -0,0 +1,171 @@
+     * 'encryption' - (string) Optionally supply an encryption or hashing
+     * 'show_encryption' - (boolean) prepend encryption info to password string
+     * </pre>
+     */
+    public function __construct(array $params = array())
+    {
+        $params = array_merge(
+            array(
+                'encryption' => 'plain',
+                'users' => array(),
+                'show_encryption' => false
+            ),
+            $params
+        );
+        parent::__construct($params);
+    }
+
+    /**
+     * Adds a set of authentication credentials.
+     *
+     * @param string $userId      The userId to add.
+     * @param array $credentials  The credentials to use.
+     *
+     * @throws Horde_Auth_Exception
+     */
+    public function addUser($userId, $credentials)
+    {
+        // TODO: use persistence callback if given
+        if ($this->exists($userId)) {
+            throw new Horde_Auth_Exception('User already exists');
+        }
+        $this->_params['users'][$userId] = Horde_Auth::getCryptedPassword(
+               $credentials['password'],
+               '',
+               $this->_params['encryption'],
+               $this->_params['show_encryption']);
+    }
+
+    /**
+     * Lists all users in the system.
+     *
+     * @param boolean $sort  Sort the users?
+     *
+     * @return mixed  The array of userIds.
+     * @throws Horde_Auth_Exception
+     */
+    public function listUsers($sort = false)
+    {
+        return $this->_sort(array_keys($this->_params['users']), $sort);
+    }
+
+    /**
+     * Updates a set of authentication credentials for the life time of the driver.
+     *
+     * @param string $oldID       The old userId.
+     * @param string $newID       The new userId.
+     * @param array $credentials  The new credentials
+     *
+     * @throws Horde_Auth_Exception
+     */
+    public function updateUser($oldID, $newID, $credentials)
+    {
+        if (!$this->exists($oldID)) {
+            throw new Horde_Auth_Exception('User does not exist');
+        }
+        if ($this->exists($newID) && $newID != $oldID) {
+            throw new Horde_Auth_Exception('Cannot rename to existing user name');
+        }
+        $this->removeUser($oldID);
+        $this->addUser($newID, $credentials);
+    }
+
+    /**
+     * Deletes a set of authentication credentials for the life of the driver.
+     *
+     * @param string $userId  The userId to delete.
+     *
+     * @throws Horde_Auth_Exception
+     */
+    public function removeUser($userId)
+    {
+        // TODO: use persistence callback if given
+        unset($this->_params['users'][$userId]);
+    }
+
+    /**
+     * Authenticate.
+     *
+     * @param string $userId      The userID to check.
+     * @param array $credentials  An array of login credentials.
+     *
+     * @throws Horde_Auth_Exception
+     */
+    protected function _authenticate($userId, $credentials)
+    {
+        if ($this->exists($userId)) {
```

Comment:
This is easier to follow by negating the check, i.e. `if (!$this->exists || $this->_compare) throw`